src/xhr-base.js

```diff
@@ -0,0 +1,404 @@
+    if (!this.ampdocSingle_) {
+      return Promise.resolve();
+    }
+    const viewer = Services.viewerForDoc(this.ampdocSingle_);
+    const whenFirstVisible = viewer.whenFirstVisible();
+    if (isProxyOrigin(input) || !viewer.hasCapability('xhrInterceptor')) {
+      return whenFirstVisible;
+    }
+    const htmlElement = this.ampdocSingle_.getRootNode().documentElement;
+    const docOptedIn = htmlElement.hasAttribute('allow-xhr-interception');
+    const isDevMode = getMode(this.win).development;
+    if (!docOptedIn) {
+      return whenFirstVisible;
+    }
+    return whenFirstVisible.then(() => {
+      return viewer.isTrustedViewer();
+    }).then(viewerTrusted => {
+      if (!viewerTrusted && !isDevMode) {
+        return;
+      }
+      const messagePayload = dict({
+        'originalRequest': this.toStructuredCloneable_(input, init),
+      });
+      return viewer.sendMessageAwaitResponse('xhr', messagePayload)
+          .then(response =>
+            this.fromStructuredCloneable_(response, init.responseType));
+    });
+  }
+
+  /**
+   * De-serializes a fetch response that was made possible to be passed to
+   * `postMessage()`, i.e., can be cloned using the
+   * [structured clone algorithm](http://mdn.io/Structured_clone_algorithm).
+   *
+   * The response is assumed to be serialized in the following way:
+   *
+   * 1. Transform the entries in the headers of the response into an
+   * `!Array<!Array<string>>` holding the list of header entries, where each
+   * element in the array is a header entry (key-value pair) represented as a
+   * 2-element array. The header key is case-insensitive.
+   *
+   * 2. Include the header entry list and `status` and `statusText` properties
+   * of the response in as `headers`, `status` and `statusText` properties of
+   * `init`.
+   *
+   * 3. Include the body of the response serialized as string in `body`.
+   *
+   * 4. Return a new object having properties `body` and `init`.
+   *
+   * The response is de-serialized in the following way:
+   *
+   * 1. If the `Response` type is supported and `responseType` is not
+   * document, pass `body` and `init` directly to the constructor of `Response`.
+   *
+   * 2. Otherwise, populate a fake XHR object to pass to `FetchResponse` as if
+   * the response is returned by the fetch polyfill.
+   *
+   * 3. If `responseType` is `document`, also parse the body and populate
+   * `responseXML` as a `Document` type.
+   *
+   * @param {JsonObject|string|undefined} response The structurally-cloneable
+   *     response to convert back to a regular Response.
+   * @param {string|undefined} unusedResponseType The original response type used to
+   *     initiate the XHR.
+   * @return {*} The deserialized regular response.
+   * @protected
+   */
+  fromStructuredCloneable_(response, unusedResponseType) {
+    user().assert(isObject(response), 'Object expected: %s', response);
+    return new this.win.Response(response['body'], response['init']);
```

Comment:
Where is `Response` being polyfilled?

Reply:
Actually its not, and hence the `fromStructuredCloneable_` for `xhr-impl` is kept as in as an override.
The polyfill will come in next PR. But i'll add a `dev().assert` in this PR so that this one doesn't get used.